xc/xenops_server_xen.ml

```diff
@@ -3333,27 +3333,45 @@ module Actions = struct
     let vm = get_uuid ~xc domid |> Uuidm.to_string in
     Opt.iter
       (function { VmExtra.persistent } ->
-         if not persistent.VmExtra.pv_drivers_detected then begin
-           (* If the new value for this device is 4 then PV drivers are present *)
-           try
-             let value = xs.Xs.read path in
-             if value = "4" (* connected *) then begin
-               let updated =
-                 DB.update vm (
-                   Opt.map (function { VmExtra.persistent } ->
-                       let persistent = { persistent with VmExtra.pv_drivers_detected = true } in
-                       debug "VM = %s; found PV driver evidence on %s (value = %s)" vm path value;
-                       VmExtra.{persistent}
-                     )
-                 )
-               in
-               if updated then
-                 Updates.add (Dynamic.Vm vm) internal_updates
-             end
-           with Xs_protocol.Enoent _ ->
-             warn "Watch event on %s fired but couldn't read from it" path;
-             () (* the path must have disappeared immediately after the watch fired. Let's treat this as if we never saw it. *)
-         end
+        try
+          let value = xs.Xs.read path in
+          let pv_drivers_detected =
+            if value = "4" && not persistent.VmExtra.pv_drivers_detected then begin
```

Comment:
you could convert `value` to a value of type `Xenbus_utils.state`

Reply:
Rather than converting the read number each time, I'll put the following on top of the function and use that in the comparisons rather than `"4"`.

    let connected = Xenbus_utils.(int_of Connected) in